### PR TITLE
DTM0: [EOS-18924] Add code to dtm0 fop handler for handling PERSISTENT

### PR DESCRIPTION
**Pull request by mehjoshi**
This PR has changes for handling persistent message handling for dtm0

### CODE REVIEW

File: dtm0/fop.c [hunk 1]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+rep = m0_fop_data(fom->fo_rep_fop);
+M0_SET0(&rep->dr_txr);
+svc = m0_dtm0_service_find(fom->fo_service->rs_reqh);
+M0_ASSERT(svc != NULL);
+
+M0_LOG(M0_DEBUG, "Processing non-generic phase %" PRIu32
+" %d, " FID_F ", %p", req->dtr_msg,
+!!m0_dtm0_is_a_volatile_dtm(fom->fo_service),
+FID_P(&fom->fo_service->rs_service_fid),
+			fom->fo_service->rs_reqh);
```
Comment (just-now) [inline dtm0/fop.c:403]: Looks like indentation has to be fixed.

File: dtm0/fop.c [hunk 2]
```diff
@@ -108,10 +108,16 @@ M0_INTERNAL void m0_dtm0_fop_fini(void)
 }
 
 enum {
```
Comment (just-now) [inline dtm0/fop.c:110]: Have UTs and ./all2all passed?

File: dtm0/fop.c [hunk 3]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+" %d, " FID_F ", %p", req->dtr_msg,
+!!m0_dtm0_is_a_volatile_dtm(fom->fo_service),
+FID_P(&fom->fo_service->rs_service_fid),
+			fom->fo_service->rs_reqh);
+
+			if (req->dtr_msg == DMT_EXECUTE &&
+					m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
+				M0_ASSERT(m0_dtm0_in_ut());
+				rc = m0_dtm0_tx_desc_copy(&req->dtr_txr, &rep->dr_txr);
+				if (rc != 0) {
```
Comment (just-now) [inline dtm0/fop.c:409]: remove unneeded {}

File: dtm0/fop.c [hunk 4]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+					m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
+				}
+				m0_dtm0_send_msg(fom, DMT_EXECUTED,
+								 &req->dtr_txr.dtd_id.dti_fid,
+								 &req->dtr_txr);
+			} else if (req->dtr_msg == DTM_PERSISTENT) {
+				if (m0_dtm0_is_a_volatile_dtm(fom->fo_service)) {
+								m0_be_dtm0_log_pmsg_post(svc->dos_log, fom->fo_fop);
```
Comment (just-now) [inline dtm0/fop.c:417]: indentation

File: dtm0/fop.c [hunk 5]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+			break;
+
+		default:
+			M0_IMPOSSIBLE("Invalid phase");
 	}
 
+	phase_sm_id = m0_sm_id_get(&fom->fo_sm_phase);
+	rpc_sm_id   = m0_sm_id_get(&item->ri_sm);
+	M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, phase_sm_id, rpc_sm_id);
 	return M0_RC(rc);
```
Comment (just-now) [inline dtm0/fop.c:438]: does it always return M0_FSO_WAIT/AGAIN?
For example, after line 391 there's no rc=AGAIN assignment.

File: dtm0/fop.c [hunk 6]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+
+			m0_be_dtm0_log_credit(M0_DTML_PERSISTENT,
+								  &txd,
+								  &buf,
+								  m0_fom_reqh(fom)->rh_beseg,
+								  NULL,
+								  &dtm0logrec_cred);
+			m0_be_tx_credit_add(&fom->fo_tx.tx_betx_cred, &dtm0logrec_cred);
+			rc = m0_be_tx_open_sync(&fom->fo_tx.tx_betx);
+			if (rc != 0) {
```
Comment (just-now) [inline dtm0/fop.c:386]: remove unneeded {}

File: dtm0/fop.c [hunk 7]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
-			 */
-			m0_be_dtm0_log_pmsg_post(svc->dos_log, fom->fo_fop);
+			m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
+			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr, &txd);
+			if (rc != 0) {
+				m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
+			}
+
+			m0_be_dtm0_log_credit(M0_DTML_PERSISTENT,
+								  &txd,
```
Comment (just-now) [inline dtm0/fop.c:379]: indentation?

**Review by just-now — COMMENTED** [review]
Overall logic and refactoring look valid.
Can be landed after the following checks:
 - ut && ./all2all.
 - fixing comments.

File: dtm0/fop.c (hunk 2 above)
Comment (mehjoshi) [reply dtm0/fop.c:110]: Existing UT has passed. I will run all2all test and update.

File: dtm0/fop.c (hunk 7 above)
Comment (mehjoshi) [reply dtm0/fop.c:379]: indentation looks different in the editor versus her in github.

File: dtm0/fop.c (hunk 1 above)
Comment (mehjoshi) [reply dtm0/fop.c:403]: OK.

File: dtm0/fop.c (hunk 5 above)
Comment (mehjoshi) [reply dtm0/fop.c:438]: Thanks. Fixed.

File: dtm0/fop.c (hunk 4 above)
Comment (mehjoshi) [reply dtm0/fop.c:417]: Done.

File: dtm0/fop.c (hunk 3 above)
Comment (mehjoshi) [reply dtm0/fop.c:409]: Done.

File: dtm0/fop.c (hunk 6 above)
Comment (mehjoshi) [reply dtm0/fop.c:386]: Done.

File: dtm0/fop.c [hunk 8]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+								  &buf,
+								  m0_fom_reqh(fom)->rh_beseg,
+								  NULL,
+								  &dtm0logrec_cred);
+			m0_be_tx_credit_add(&fom->fo_tx.tx_betx_cred, &dtm0logrec_cred);
+			rc = m0_be_tx_open_sync(&fom->fo_tx.tx_betx);
```
Comment (ivan-alekhin) [inline dtm0/fop.c:385]: Why do we need to explicitly open_sync/close_sync this tx?  Wouldn't it be done as a part of the generic phases?

Example:

```
case <generic-phases>:
    rc = tick_generic(fom);
    next_phase = fom_phase(fom);
    /* Calculate the credit before opening the tx */
    if (next_phase == TX_OPEN)
        caclulate_log_record_credits(...);
    break;
```

File: dtm0/fop.c [hunk 9]
```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
...
+			}
 		}
-
-		rep->dr_rc = 0;
-		m0_fom_phase_set(fom, M0_FOPH_SUCCESS);
-		rc = M0_FSO_AGAIN;
+		m0_fom_phase_set(fom, M0_FOPH_DTM0_LOGGING);
+		break;
+
+case M0_FOPH_DTM0_LOGGING:
```
Comment (ivan-alekhin) [inline dtm0/fop.c:393]: Let's fix indentation here.

**Review by ivan-alekhin — COMMENTED** [review]
A couple of questions from my side.

File: dtm0/fop.c (hunk 8 above)
Comment (mehjoshi) [reply dtm0/fop.c]: Applied the changes as discussed.

File: dtm0/fop.c [hunk 10]
```diff
@@ -343,19 +343,61 @@ M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom                *fom,
...
+				m0_be_dtm0_log_credit(M0_DTML_PERSISTENT,
+						      &req->dtr_txr,
+						      &buf,
+						      m0_fom_reqh(fom)->rh_beseg,
+						      NULL,
+						      &dtm0logrec_cred);
+				m0_be_tx_credit_add(&fom->fo_tx.tx_betx_cred, &dtm0logrec_cred);
+			}
+			if (phase == M0_FOPH_TXN_COMMIT) {
+				rc = op_sync_wait(fom);
```
Comment (ivan-alekhin) [inline dtm0/fop.c:393]: I'm not sure if we really need to wait here until the tx gets logged. Let's discuss this.

File: dtm0/fop.c [hunk 11]
```diff
@@ -343,19 +343,61 @@ M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom                *fom,
...
+			}
+			if (m0_dtm0_in_ut() && m0_fom_phase(fom) == M0_FOPH_QUEUE_REPLY) {
+				m0_fom_phase_set(fom, M0_FOPH_TXN_COMMIT_WAIT);
```
Comment (ivan-alekhin) [inline dtm0/fop.c:396]: Why do we skip `M0_FOPH_QUEUE_REPLY` phase in the UT?

File: dtm0/fop.c (hunk 10 above)
Comment (mehjoshi) [reply dtm0/fop.c]: Done. Removed this code after discussion with Ivan.

File: dtm0/fop.c (hunk 11 above)
Comment (mehjoshi) [reply dtm0/fop.c]: Done. Updated code after discussion with Ivan.

File: dtm0/fop.c (hunk 2 above)
Comment (mehjoshi) [reply dtm0/fop.c]: I have updated the UT for dtm0 client to send and validate persistent messages. I have tested with these UT changes.

File: dtm0/fop.c [hunk 12]
```diff
@@ -369,28 +392,36 @@ static int dtm0_fom_tick(struct m0_fom *fom)
...
-						  &rep->dr_txr);
-			M0_ASSERT(rc == 0);
+			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr, &rep->dr_txr);
+			if (rc != 0)
+				m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
 			rc = m0_dtm0_send_msg(fom, DMT_EXECUTED,
 					      &req->dtr_txr.dtd_id.dti_fid,
 					      &req->dtr_txr);
-			M0_ASSERT(rc == 0);
+			if (rc != 0)
```
Comment (ivan-alekhin) [inline dtm0/fop.c:401]: Some of these line are longer than 80 characters. Let's try to see if we can fix it. If looks like we too many levels of nested if/else-s here, I think we could try to reduce the amount of levels to improve readability of the code and to make this code aligned with the coding style.

File: dtm0/ut/main.c [hunk 13]
```diff
@@ -109,7 +116,41 @@ static void dtm0_ut_send_fops(struct m0_rpc_session *cl_rpc_session)
 	M0_ASSERT(m0_dtm0_ts__invariant(&reply_data.dti_ts));
 
 	M0_UT_ASSERT(m0_dtm0_tid_cmp(&dcs, &txr.dtd_id, &reply_data) == M0_DTS_EQ);
+	m0_fop_put_lock(fop);
```
Comment (ivan-alekhin) [inline dtm0/ut/main.c:119]: Let's see if we can move this code (send P msg, recv it, send P msg back) into its own test case. This is optional, we can address it later.

**Review by ivan-alekhin — COMMENTED** [review]
The approach looks fine to me. I ran the test and checked the sequence, it also looks fine.

Let's see if we can address these coding style-related comments before merging it into dtm0-main.

File: dtm0/fop.c (hunk 12 above)
Comment (mehjoshi) [reply dtm0/fop.c:401]: OK. Will do.

File: dtm0/ut/main.c (hunk 13 above)
Comment (mehjoshi) [reply dtm0/ut/main.c:119]: Yes, I will make a note in the code and address it later.

**Comment by mehjoshi** [comment]
Upon inspecting the changes again I realized that the changes are incomplete or some portions of code changes are missing.
Will redo all the changes with the latest branch and open another PR post testing

**Pull request closed without merging**